### PR TITLE
Hund radio system

### PR DESCRIPTION
**Pull request by NikkoJT**
This PR includes the new ingame long-range radio module based on @Hundenal 's prototype.

The module creates custom radio channels, and authorises access based on whether a player has an eligible inventory item, or is in an eligible radio vehicle, or has a radio channel set by variable. Inventory items, variables, and being a vehicle driver or commander give talk access, while being in a non-driver vehicle position gives listen-only access.

There are two modes: unified mode, where all eligible radio items give access to one channel, and split mode, where each of the 10 channels has a set of eligible items.

The module uses event handlers to update access whenever a player puts or takes an item, and enters, exits, or changes seats in a vehicle.

Issue #357

### CODE REVIEW

File: f/radio/f_radioChannels.sqf [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	publicVariable "f_var_radioChannelsBackpacks";
+	publicVariable "f_var_radioChannelsVehicles";
+	publicVariable "f_var_radioChannelsUnified";
+	
+	// Add server EH for JIP
+	addMissionEventHandler ["PlayerConnected",
+	{
+		params ["_id", "_uid", "_name", "_jip", "_owner", "_idstr"];
+		if (_jip) then {
+			[] remoteExec ["f_fnc_radioAddHandlers",_owner];
```
Comment (shadow-fa) [inline f/radio/f_radioChannels.sqf:114]: Doesn't this need a parameter `_splitMode`?

Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:118]: Addressed by reworking the splitmode switch into a global variable. As a side effect this means the mode can be changed mid-mission if required.

File: f/radio/fn_radioAddHandlers.sqf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// Update channels if they switch seats in a vehicle
+player addEventHandler ["seatSwitchedMan", {  
+	[_splitMode] call f_fnc_radioCheckChannels;  
+}];
+
+// Just to be sure...
+2 enableChannel false;
+
+// Set a variable on the player to prove they've got handlers
+player setVariable ["f_var_radioHandlersAdded",true,true];
```
Comment (shadow-fa) [inline f/radio/fn_radioAddHandlers.sqf:47]: Strictly speaking, this doesn't need to be broadcast

Comment (NikkoJT) [reply f/radio/fn_radioAddHandlers.sqf]: Removed it - it could actually have caused problems if the unit changed locality for some reason, because the variable would transfer but not the actual handlers

File: f/radio/fn_radioAddHandlers.sqf [hunk 3]
```diff
@@ -0,0 +1,47 @@
+// F3 - Folk ARPS Long-Range Radio Module
+// Credits and documentation: https://github.com/folkarps/F3/wiki
+
+/* ========================
+This function adds local event handlers to the player which grant radio channels when picking up a backpack or entering a vehicle. It also makes an initial check to see what they've got to begin with.
+It's activated by f\radio\f_radioChannels.sqf.
+=========================== */
+waitUntil {!(isNil "f_var_radioChannelsUnified")};
+
+_splitMode = f_var_radioSplitMode;
```
Comment (shadow-fa) [inline f/radio/fn_radioAddHandlers.sqf:10]: I'm not sure if there's a guaranteed order of broadcast variables. So depending on how much faith we have in arma we could also wait for the `f_var_radioSplitMode` variable.

Comment (shadow-fa) [reply f/radio/fn_radioAddHandlers.sqf:10]: also, `private`. and variable isn't even used?

File: f/radio/fn_radioCheckChannels.sqf [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+=========================== */
+
+// Clients wait for the server to finish with setup
+waitUntil {!(isNil "f_var_radioChannelsUnified")};
+
+_splitMode = f_var_radioSplitMode;
+
+_channelsToAddListen = [];
+_channelsToAddTalk = [];
+_isVehicle = false;
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:15]: all of the variables should probably be marked as `private`

Comment (shadow-fa) [reply f/radio/fn_radioCheckChannels.sqf:15]: `_isVehicle` isn't used?

File: f/radio/fn_radioAddHandlers.sqf (hunk 3 above)
Comment (NikkoJT) [reply f/radio/fn_radioAddHandlers.sqf:10]: On mission start, the variables should be broadcast in order of instruction (at least when there's this much other stuff happening between instructions). On JIP, all public variables are sent before any scripts are run, so it's not a problem.

File: f/radio/fn_radioCheckChannels.sqf (hunk 4 above)
Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: Fixed

Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: _isVehicle removed; was part of a different system that got halfway before being replaced.

File: f/radio/fn_radioCheckChannels.sqf [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+private _radioChannelsPlayerSpecific = [];
+private _radioChannelsVehicleSpecific = [];
+private _channelChecked = 0;
+private _channelBackpacks = [];
+private _channelsToAddListen = [];
+private _channelsToAddTalk = [];
+
+for "_i" from 1 to 10 do {
+
+	_channelChecked = _i;
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:23]: i would get rid of `_channelChecked` or `_i` and only use one of them

File: f/radio/fn_radioAddHandlers.sqf (hunk 3 above)
Comment (NikkoJT) [reply f/radio/fn_radioAddHandlers.sqf:10]: _splitMode could be made private but I don't see any particular need to. It should always be the same in all contexts because it's a global setting.

f_var_radioChannelsUnified is not currently used (would have been used in an alternate version of the mode system) but I think it is potentially useful to expose for missionmakers.

File: f/radio/fn_radioCheckChannels.sqf [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	if ((backpack player) in _channelBackpacks) then {
+		_channelsToAddListen pushBackUnique _channelChecked;
+		_channelsToAddTalk pushBackUnique _channelChecked;
+	};
+	
+	// Check for vehicles. Don't add send permissions unless they're the driver.
+	_channelVehicles = (f_var_radioChannelsVehicles get _i);
+	if ((vehicle player) in _channelVehicles) then {
+		_channelsToAddListen pushBackUnique _channelChecked;
+		if (player == driver vehicle player) then {
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:33]: does `driver` return `objNull` or something else if there is no driver? I think `isEqualTo` is suggested for these types of comparisons

File: f/radio/fn_radioCheckChannels.sqf (hunk 5 above)
Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: Fixed

File: f/radio/fn_radioCheckChannels.sqf (hunk 6 above)
Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf:33]: `driver` returns `objNull` so this is safe

File: f/radio/f_radioChannels.sqf [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	// Backpacks for channel 7
+	f_var_radioChannelsBackpacks set [7, ["B_RadioBag_01_mtp_F"]];
+	// Backpacks for channel 8
+	f_var_radioChannelsBackpacks set [8, ["B_RadioBag_01_wdl_F"]];
+	// Backpacks for channel 9
+	f_var_radioChannelsBackpacks set [9, ["B_RadioBag_01_oucamo_F"]];
+	// Backpacks for channel 10
+	f_var_radioChannelsBackpacks set [10, []];
+
+	// VEHICLES: Can be a variable name of a specific vehicle (VARIABLE), or a classname of a vehicle type (STRING)
```
Comment (shadow-fa) [inline f/radio/f_radioChannels.sqf:77]: > Can be a variable name of a specific vehicle

Was this tested? Can we add an example?
We might need to use strings of variable names and then something like `missionNamespace getVariable [_someString, objNull]`

File: f/radio/fn_radioCheckChannels.sqf [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+_radioChannelsPlayerSpecific = (player getVariable ["f_var_radioChannelsObjectSpecific",[]]);
+if ((count _radioChannelsPlayerSpecific) > 0) then {
+	{
+		_channelsToAddListen pushBackUnique _x;
+		_channelsToAddTalk pushBackUnique _x;
+	} forEach _radioChannelsObjectSpecific;
+};
+
+_radioChannelsVehicleSpecific = (vehicle player getVariable ["f_var_radioChannelsObjectSpecific",[]])
+if ((count _radioChannelsPlayerSpecific) > 0) then {
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:56]: should this be `_radioChannelsVehicleSpecific`?

File: f/radio/fn_radioCheckChannels.sqf [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	};
+};
+
+// Detect any channels activated by setting a variable on the player or their vehicle
+_radioChannelsPlayerSpecific = (player getVariable ["f_var_radioChannelsObjectSpecific",[]]);
+if ((count _radioChannelsPlayerSpecific) > 0) then {
+	{
+		_channelsToAddListen pushBackUnique _x;
+		_channelsToAddTalk pushBackUnique _x;
+	} forEach _radioChannelsObjectSpecific;
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:52]: should this be `_radioChannelsPlayerSpecific`?

Comment (shadow-fa) [reply f/radio/fn_radioCheckChannels.sqf:52]: maybe removing the `if` and the variable name is less prone to errors?

File: f/radio/fn_radioCheckChannels.sqf [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+_radioChannelsVehicleSpecific = (vehicle player getVariable ["f_var_radioChannelsObjectSpecific",[]])
+if ((count _radioChannelsPlayerSpecific) > 0) then {
+	{
+		_channelsToAddListen pushBackUnique _x;
+		if (player == driver vehicle player) then {
+			_channelsToAddTalk pushBackUnique _x;
+		};
+	} forEach _radioChannelsObjectSpecific;
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf]: same here

File: f/radio/fn_radioCheckChannels.sqf (hunk 8 above)
Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: Fixed by removal

File: f/radio/fn_radioCheckChannels.sqf (hunk 10 above)
Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: Done

File: f/radio/fn_radioAddHandlers.sqf (hunk 3 above)
Comment (shadow-fa) [reply f/radio/fn_radioAddHandlers.sqf:10]: What I meant was that the variable `_splitMode` is never actually used anywhere in this file.

Comment (shadow-fa) [reply f/radio/fn_radioAddHandlers.sqf:10]: I don't see why they wouldn't just use `f_var_radioSplitMode`

File: f/radio/f_radioChannels.sqf (hunk 7 above)
Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf]: Fixed by stringification of vehicles in definition and checking

File: f/radio/fn_radioCheckChannels.sqf [hunk 11]
```diff
@@ -28,7 +28,7 @@ for "_i" from 1 to 10 do {
 	
 	// Check for vehicles. Don't add send permissions unless they're the driver.
 	_channelVehicles = (f_var_radioChannelsVehicles get _i);
-	if ((vehicle player) in _channelVehicles) then {
+	if ((str vehicle player) in _channelVehicles) then {
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf:31]: I'm not certain `str` always returns the variable name of an object. is this documented somewhere? It's possible that `vehicleVarName` can be used.

Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf:31]: If a vehicle has a variable name, str returns it; if it doesn't, it returns an identifier something like `B Alpha 1-1:1 NikkoJT`, which is fine since we shouldn't be matched it if it has no variable name. I tested it.

Comment (shadow-fa) [reply f/radio/fn_radioCheckChannels.sqf:31]: Oh of course it was only "documented" in the [community-provided notes](https://community.bistudio.com/wiki/str#Notes): `When applied to a unit, this returns the variable name that was assigned to the unit in the editor.`

**Comment by shadow-fa** [comment]
Is there a reason why we're separating backpacks and vehicles? Couldn't the same array hold both backpacks and vehicles?

**Comment by NikkoJT** [comment]
I think originally there was some technical reason. Now it seems like it may be technically practical to have them both in one array.

That being said, there could also be a reasonable benefit to readability by keeping them separate, since the arrays must be filled out by the missionmaker.

File: f/radio/f_radioChannels.sqf [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+Example:
+[true] execVM "f\radio\f_radioChannels.sqf";
+
+Arguments:
+0. Long-range split mode (boolean). If true, each radio backpack type has a separate channel. If false, all backpacks use one channel. Default false.
+
+This module requires some missionmaker configuration. If you want to enable vehicle radio access, you need to specify vehicle classes or variable names below.
+=========================== */
+
+params [["_splitMode",false]];
```
[the remark this answers is not in the document]
Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:15]: It was more important in the first iteration of the system (before the vehicle radio refactor) when the channels were generated differently. So in some ways it's a holdover. However, I think having the ability to use a simple toggle rather than futzing with arrays can be useful. The radio system does work in the default unconfigured state, using only backpacks and no vehicles, so if that's all that's required for a mission, setting the split mode is all they need to do to be up and running. It also provides missionmakers the ability to quickly reuse a set of eligible items in a different mode, without needing to change the arrays, as well as a simple modular hook to connect to e.g. a mission parameter.

File: f/radio/f_radioChannels.sqf [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+*/
+
+if (isServer) then {
+
+	// Make the split mode something we can reference later
+	f_var_radioSplitMode = _splitMode;
+	publicVariable "f_var_radioSplitMode";
+
+	// This will be used later
+	f_var_radioChannelUnified = [];
```
Comment (shadow-fa) [inline f/radio/f_radioChannels.sqf:36]: I'm posting this here to keep it related answers together in the same thread:

I suggest merging the 3 arrays (channel names, backpacks, vehicles) into a singular simple structure:
```sqf
f_var_radioChannels = [
	// Arrays consist of channel name (as shown ingame) and a list of backpacks, vehicle types, or vechile variable names (as strings)
	["LR Channel 1",   ["B_SDV_01_F", "B_RadioBag_01_black_F"]],
	["LR Specialists", ["C_Tractor_01_F", "B_RadioBag_01_mtp_F"]],
];
```

I think it helps if things that belong in the same channel are also specified in the same place

Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:36]: At the moment the arrays of eligible classnames are actually Hashmaps of arrays, not just arrays. This is an important distinction because the hashmap key for each channel is also the channel ID, and this is cross-referenced and used in the channel add/remove function. I seriously doubt whether the ID key can be replaced by a customisable channel name in this usage, or whether moving to arrays instead of hashmaps is likely to be an improvement to complexity or performance.

The hashmaps of vehicles and backpacks could be merged easily.

Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:36]: I wasn't implying that we should move away from the hashmaps. The above array can be converted (at runtime) to a hashmap easily (if it turns out that channel numbers are different than e.g., `_forEachIndex + 1`).

File: f/radio/f_radioChannels.sqf [hunk 14]
```diff
@@ -56,72 +56,50 @@ if (isServer) then {
...
+	f_var_radioChannelsObjects set [10, []];
 	
 	// You can also tag a specific unit or vehicle for access to specific channels by setting a variable on them:
 	// _unit setVariable ["f_var_radioChannelsObjectSpecific",[1,2,3],true];
 
 	// Flatten all these arrays into one single list for potential later use
-	{
-		f_var_radioChannelsUnified append (flatten (values _x));
-	} forEach [f_var_radioChannelsBackpacks,f_var_radioChannelsVehicles];
+	f_var_radioChannelsUnified append (flatten (values f_var_radioChannelsObjects));
```
Comment (shadow-fa) [inline f/radio/f_radioChannels.sqf:98]: do we still need this?

File: f/radio/f_radioChannels.sqf (hunk 13 above)
Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:36]: The default configuration should accommodate all template factions, in split mode, which means we need a channel for all 9 vanilla backpacks, which is near to 10 as makes no odds. The alternative is to go back to having duplicate generation methods based on which mode we're running in, which I'm not a fan of.

File: f/radio/f_radioChannels.sqf (hunk 14 above)
Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:98]: > f_var_radioChannelsUnified is not currently used (would have been used in an alternate version of the mode system) but I think it is potentially useful to expose for missionmakers.

This still applies. f_var_radioChannelsUnified is still meaningfully different to f_var_radioChannelsObjects - it's now an array rather than a hashmap, and it lists all of the radio objects without any of the keys so they can be easily accessed for any other purpose.

Will have to adjust how it's generated if switching to a single hashmap for all channel properties.

Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:98]: > Will have to adjust how it's generated if switching to a single hashmap for all channel properties.

probably something like 
```sqf
f_var_radioChannelsUnified append (flatten (values f_var_radioChannelsObjects apply {_x select 1}));
```
and maybe `f_var_radioChannelsUnified = f_var_radioChannelsUnified arrayIntersect f_var_radioChannelsUnified;` for deduplication.

File: f/radio/f_radioChannels.sqf (hunk 13 above)
Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:36]: I agree that the default should accommodate everything. But right now it is hard to use custom channels (for other purposes) when this module is active. The added code complexity is worth it to me for the increased compatibility and ease of use

Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:36]: This module currently assumes that no other code (which may run at the same time) calls radioChannelCreate because we don't properly check/use the return values.

Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:41]: I think I found a semi-elegant solution

Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:41]: I'm not sure it has to be a parameter since we can just take the information from the `f_var_radioChannels`.

I'm not a fan of having the channel IDs in the same structure that the missionmaker modifies. If a missionmaker comments out one of these lines `f_var_radioChannels set [...];` then the channel numbers may be mismatched compared to the ones that we actually create. 
If we keep it like this, we want to have checks in place to prevent these issues. That is, IDs returned from `radioChannelCreate` must match the keys in `f_var_radioChannels`. This should be as simple as comparing `_ChannelID` with `_i + some_offset`.

Comment (shadow-fa) [reply f/radio/f_radioChannels.sqf:41]: We could also print some warning/error if the provided number (`_channelCount`) is smaller than the non-empty entries in `f_var_radioChannels `.

Comment (NikkoJT) [reply f/radio/f_radioChannels.sqf:41]: A missionmaker commenting out random lines from the middle of an F3 script can break pretty much anything. There is no ultimate safety against code destruction.
If you want to put the brainpower into figuring out how to make that work then go for it, but I don't.

File: f/radio/f_radioChannels.sqf [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	// If channels are not to be split, only create one.
+	if (!_splitMode) then {
+		_channelCount = 1;
+	};
+	
+	// Set up channels to use
+	for "_i" from 0 to (_channelCount - 1) do {
+		_ChannelName = format ["%1",((_channelNameList get _i) select 0)];
+		_ChannelID = (radioChannelCreate [[0.96, 0.34, 0.13, 0.8], _ChannelName, "%UNIT_NAME", []]);
+		if (_ChannelID == 0) exitWith {diag_log format ["channel creation failed", _x]};
```
Comment (shadow-fa) [inline f/radio/f_radioChannels.sqf:86]:
```suggestion
		if (_ChannelID != _i + 1) exitWith {diag_log format ["channel creation failed", _x]};
```
This should make it safer. (untested)

File: f/radio/fn_radioCheckChannels.sqf [hunk 16]
```diff
@@ -5,9 +5,6 @@
 This function checks what channels a player should have access to. It's called by event handlers set up by fn_radioAddHandlers.sqf.
 =========================== */
 
-// Clients wait for the server to finish with setup
-waitUntil {!(isNil "f_var_radioChannelsUnified")};
```
Comment (shadow-fa) [inline f/radio/fn_radioCheckChannels.sqf]: Why was this removed? I think we do need to wait for the server to initialize and broadcast all the `f_var_radio` variables here

Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: The server broadcasts the publicVariable before remoteExec'ing the next function to the client. According to Dedmen, those messages are guaranteed to be handled in order of transmission, so the client will always have the variables by the time it runs the function.

Comment (shadow-fa) [reply f/radio/fn_radioCheckChannels.sqf]: I'm not sure I understand.
Here, the client calls this function directly. There is no remoteExec involved:
https://github.com/folkarps/F3/blob/72f33d443dec5fc1b9ae18359192265c5caaf311/f/radio/f_radioChannels.sqf#L106

Comment (NikkoJT) [reply f/radio/fn_radioCheckChannels.sqf]: Confused it with something else. I've replaced that waitUntil as required, but it is now located in f_fnc_radioAddHandlers. It can't be located in f_fnc_radioCheckChannels because that function can be unscheduled, where waitUntil doesn't work properly.

**Comment by NikkoJT** [comment]
Current status: this system works as expected - //when it works//. The one remaining issue is that //occasionally// a change in status (e.g. getting out of a vehicle or changing radio pack) will not be detected. There doesn't seem to be a consistent cause for this, and I can't tell whether it's the EH not firing or something going wrong in the channel updater.

Live testing is ongoing - the issue can be worked around by just triggering another check by doing an inventory movement or whatever - but I don't want to release this as stock F3 until this can be resolved. Unfortunately, I'm not getting any closer to resolving it  as I have no idea what's happening. Any insights would be useful.

**Comment by SamLex** [comment]
@NikkoJT is this ready for review? If so, could you remove the WIP label and unset as a draft PR please?

**Comment by NikkoJT** [comment]
Not quite. I'm waiting for another mission or two in which we use the latest version, in order to prove the reliability rate of the double-check system.

File: f/radio/f_channelsList.sqf [hunk 17]
```diff
@@ -0,0 +1,230 @@
+// F3 - Folk ARPS Long-Range Radio Module
```
Comment (SamLex) [inline f/radio/f_channelsList.sqf:1]: If we weren't so close to release, I'd suggest moving this to config (`description.ext` et al.). I'll add it as a future idea instead.

File: f/radio/fn_radioChannels.sqf [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+0. Long-range split mode (boolean). If true, each radio backpack type has a separate channel. If false, all backpacks use one channel. Default false.
+1. Number of channels to create (number, optional). If this is set, the number of channels to be created will be limited to this number. If channels are not split, this will be automatically set to 1.
+
+This module requires some missionmaker configuration. If you want to enable vehicle radio access, you need to specify vehicle classes or variable names in "f\radio\f_channelsList.sqf".
+=========================== */
+
+params [["_splitMode",false],["_channelCount",10]];
+
+// Disable the Command channel to avoid cheating
+2 enableChannel false;
```
Comment (SamLex) [inline f/radio/fn_radioChannels.sqf:19]: Do we also want to (ensure) other channels like global and side are disabled as well?

File: f/radio/fn_radioChannels.sqf [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+	publicVariable "f_var_radioChannelCount";
+
+	// This will be used later
+	f_var_radioChannelUnified = [];
+	
+	// Define the list of channel properties. Go to this file to change which items and vehicles grant radio access.
+	#include "f_channelsList.sqf";
+	
+	f_var_radioChannels = createHashmap;
+	f_var_radioChannels set [1, [_channelName1,_channelColour1, (_channelList1 apply {toLower _x})]];
```
Comment (SamLex) [inline f/radio/fn_radioChannels.sqf:38]: The programmer in me wants to suggest making this a loop with `compile` but I'll resist :joy:

File: f/radio/fn_radioChannels.sqf [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+	f_var_radioChannelCount = _channelCount;
+	publicVariable "f_var_radioChannelCount";
+
+	// This will be used later
+	f_var_radioChannelUnified = [];
+	
+	// Define the list of channel properties. Go to this file to change which items and vehicles grant radio access.
+	#include "f_channelsList.sqf";
+	
+	f_var_radioChannels = createHashmap;
```
Comment (SamLex) [inline f/radio/fn_radioChannels.sqf:37]: Why HashMap with number keys instead of just an array where index = number

File: f/radio/fn_radioChannels.sqf [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+	f_var_radioChannels set [4, [_channelName4,_channelColour4, (_channelList4 apply {toLower _x})]];
+	f_var_radioChannels set [5, [_channelName5,_channelColour5, (_channelList5 apply {toLower _x})]];
+	f_var_radioChannels set [6, [_channelName6,_channelColour6, (_channelList6 apply {toLower _x})]];
+	f_var_radioChannels set [7, [_channelName7,_channelColour7, (_channelList7 apply {toLower _x})]];
+	f_var_radioChannels set [8, [_channelName8,_channelColour8, (_channelList8 apply {toLower _x})]];
+	f_var_radioChannels set [9, [_channelName9,_channelColour9, (_channelList9 apply {toLower _x})]];
+	f_var_radioChannels set [10, [_channelName10,_channelColour10,(_channelList10 apply {toLower _x})]];
+	
+	// You can also tag a specific unit or vehicle for access to specific channels by setting a variable on them:
+	// _unit setVariable ["f_var_radioChannelsObjectSpecific",[1,2,3],true];
```
Comment (SamLex) [inline f/radio/fn_radioChannels.sqf:50]: Add to `init.sqf` comment?

File: f/radio/fn_radioAddHandlers.sqf [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+// Add player to the correct channels if they have a backpack
+[player] call f_fnc_radioCheckChannels;
+
+// Now bail if they've already been handled
+if (player getVariable ["f_var_radioHandlersAdded",false]) exitWith {};
+
+// Update channels if they drop a backpack
+player addEventHandler ["put", { 
+	params ["_unit", "_container", "_item"];
+	[_unit] call f_fnc_radioCheckChannels;
```
Comment (SamLex) [inline f/radio/fn_radioAddHandlers.sqf:21]: Since, as far as I can see, whenever we call `f_fnc_radioCheckChannels` we call it twice with a delay to make sure it worked, won't it be better to just add a 2 iteration loop to `f_fnc_radioCheckChannels` itself?

**Review by SamLex — DISMISSED** [review]
Looks pretty good to me. A few questions but this has been live tested enough that it should be good to go/

File: f/radio/fn_radioChannels.sqf (hunk 20 above)
Comment (NikkoJT) [reply f/radio/fn_radioChannels.sqf:37]: Hashmaps are better for performance, especially since we do a lot of lookups to this one.

File: f/radio/fn_radioChannels.sqf (hunk 18 above)
Comment (NikkoJT) [reply f/radio/fn_radioChannels.sqf:19]: I think it's best to leave those. The server already disables voice (at least for side) and the text chat is legitimately used for pre-mission banter and meta stuff. The main goal of this is a CC replacement so I don't want to intrude too far into removing commonly-used functionality outside of that.

File: f/radio/fn_radioAddHandlers.sqf (hunk 22 above)
Comment (NikkoJT) [reply f/radio/fn_radioAddHandlers.sqf:21]: Maybe, probably? I'll think about this and come back to it in the morning.

Comment (SamLex) [reply f/radio/fn_radioAddHandlers.sqf:21]: @NikkoJT What did you decide here? Personally think it'd be cleaner but happy to approve if you disagree

Comment (NikkoJT) [reply f/radio/fn_radioAddHandlers.sqf]: Loop added

Comment (SamLex) [reply f/radio/fn_radioAddHandlers.sqf]: Awesome, thanks!

**Review by SamLex — DISMISSED** [review]

File: init.sqf [hunk 23]
```diff
@@ -218,3 +218,19 @@ f_var_viewDistance_crewOnly = true;
...
+// ====================================================================================
+
+// F3 - Radio Channels
+// Credits and documentation: https://github.com/folkarps/F3/wiki
+// True to split channels by radio backpack type. Else one channel for all.
+// Optional second parameter (number): limit the channel count to this number. This frees up space for any other custom channels. Defaults to 10 (max) if not set, or 1 (minimum) if channels not split.
+// You can also tag a specific unit or vehicle for access to specific channels by setting a variable on them:
+// _unit setVariable ["f_var_radioChannelsObjectSpecific",[1,2,3],true];
+
+[false] spawn f_fnc_radioChannels;
```
Comment (SamLex) [inline init.sqf:231]: Should we add a note about note disabling this?

**Review by SamLex — DISMISSED** [review]

**Review by SamLex — APPROVED** [review]
